### PR TITLE
refactor Canvas

### PR DESCRIPTION
**Pull request by lihebi**
Refactor:
- [refactor 1] **split node types** into `nodes/[Code|Scope|Rich].tsx`.

- [refactor 2] separate logics from the **Canvas component into hooks**:
  - `useCopyPaste`: the copy paste cut logic
  - `useNodeLocation`: provide `checkNodesEndLocation` to check if the node is dropped into a scope, if so, set parent and update levels.
  - `useNodeOperations`: provide `addNode` to create new nodes
  - `useInitNodes`: the logic to load nodes on entry. Steps:
    1. connect Yjs provider
    2. If Y.doc is inconsistent with database, use database to overwrite Y.doc

- [refactor 3] cleaner separation among the following **three data stores: (1) Yjs, (2) Zustand store, (3) database**.
  - the Canvas only operates on Yjs's`nodesMap` (with a few exceptions e.g., `addPod`)
  - The nodesMap operation (`add`, `delete`, `update(x,y,width,height,parent)`)) will trigger observer to push to Zustand store and remote database. Here's the logic:
    1. Always push to Zustand store
    2. When **`transaction.local === true && !node.data?.clientId`** (i.e., the change is made by the current user, and it is not a temporary pasting node), write to database.
    - Additional note: now this `update(x,y,width,height,parent)` observer will push node position and parent to Zustand store and database, so
      - there's no need to `useEffect` on `xPos`,`yPos` in e.g. CodeNode.
      - there's no need to `setPodPosition` and `setPodParent` in `checkNodesEndLocation`

Other small refactors:
- use `node.width` instead of `node.style.width`
- use `node.data.level` instead of `node.level`
- `deletePod(apolloClient, {id, toDelete})` no longer accepts `toDelete`.

### CODE REVIEW

File: ui/src/lib/nodes.tsx [hunk 1]
```diff
@@ -128,19 +148,22 @@ export function useNodesStateSynced(nodeList) {
         } else if (change.action === "delete") {
           const node = change.oldValue;
           console.log("todelete", node);
-          deletePod(null, { id: node.id, toDelete: [] });
+          deletePod(apolloClient, { id: node.id, toDelete: [] });
```
[the remark this answers is not in the document]
Comment (lihebi) [reply ui/src/lib/nodes.tsx:151]: Yes, I’m adding a condition and setting it to null for remote peers.

File: ui/src/components/Canvas.tsx [hunk 2]
```diff
@@ -823,225 +107,330 @@ function nodetype2dbtype(nodetype: string) {
...
+      // check if the formal node is located in a scope, if it is, change its parent
+      const currentNode = reactFlowInstance.getNode(pasting);
+      if (currentNode) {
+        checkNodesEndLocation(event, [currentNode], "ROOT");
+      }
+      //clear the pasting state
+      setPasting(null);
+      // delete the original (hidden) node
+      if (cutting) {
+        nodesMap.delete(cutting);
```
[the remark this answers is not in the document]
Comment (lihebi) [reply ui/src/components/Canvas.tsx:243]: I feel `deleteElements` is the same thing as  (and [triggers](https://github.com/codepod-io/codepod/blob/98b9c5d1a642d3e7795cbdb6f7fda4c9f72f1162/ui/src/lib/nodes.tsx#L74)) `nodesMap.delete`.

[the remark this answers is not in the document]
Comment (li-xin-yi) [reply ui/src/components/Canvas.tsx:243]: Not really. [`onNodesDelete`](https://github.com/codepod-io/codepod/blob/98b9c5d1a642d3e7795cbdb6f7fda4c9f72f1162/ui/src/components/Canvas.tsx#L1196) only be called when `deleteElements` is explicitly call or nodes are deleting by shortcut (`del`). From that, we can distinguish whether the deletion is issued by local user or remote peers.

[the remark this answers is not in the document]
Comment (lihebi) [reply ui/src/components/Canvas.tsx:243]: I think `onNodesChange` covers `onNodesDelete`. I actually removed `onNodesDelete`.

I plan to use Yjs's `transaction.origin` ([ref](https://docs.yjs.dev/api/document-updates#example-building-a-custom-provider)) to distinguish between local and remote edits.

[the remark this answers is not in the document]
Comment (li-xin-yi) [reply ui/src/components/Canvas.tsx:243]: I see, I think you can call remote deletion to DB only in `onNodesChange` here https://github.com/codepod-io/codepod/blob/abc81df884096335175fb8cad84309f82f03d426/ui/src/lib/nodes.tsx#L79

Then all deletion observed from `nodesMap` are synced from remote peers (yes maybe from myself, but `deletePod` will check if the pod exists first).

Anyway, it's up to you. But be careful, I don't know if the `onNodesChange` really exactly covers only all explicit deletions. I will test it when your PR finishes tomorrow.

File: ui/src/lib/nodes.tsx [hunk 3]
```diff
@@ -35,76 +37,94 @@ export function useNodesStateSynced(nodeList) {
...
+
+  const triggerUpdate = useCallback(() => {
+    setNodes(
+      Array.from(nodesMap.values())
+        .filter(
+          (node) =>
+            !node.data.hasOwnProperty("clientId") ||
+            node.data.clientId === clientId
+        )
+        .sort((a: Node, b: Node) => a.data.level - b.data.level)
```
[the remark this answers is not in the document]
Comment (lihebi) [reply ui/src/lib/nodes.tsx:104]: Thanks, I've added this to `triggerUpdate`.

File: ui/src/components/Canvas.tsx [hunk 4]
```diff
@@ -1145,46 +506,267 @@ export function Canvas() {
...
+        // throw new Error("Inconsistent state");
+        nodes.forEach((node) => {
+          if (!nodesMap.has(node.id)) {
+            nodesMap.set(node.id, node);
+          }
+        });
+      }
+      // NOTE we have to trigger an update here, otherwise the nodes are not
+      // rendered.
+      triggerUpdate();
```
Comment (lihebi) [inline ui/src/components/Canvas.tsx:703]: This is actually where I use `triggerUpdate`. Without this, the canvas will not show anything.

File: ui/src/components/nodes/Scope.tsx [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+            <IconButton
+              size="small"
+              onClick={(e: any) => {
+                // This does not work, will throw "Parent node
+                // jqgdsz2ns6k57vich0bf not found" when deleting a scope.
+                //
+                // nodesMap.delete(id);
+                //
+                // But this works:
+                reactFlowInstance.deleteElements({ nodes: [{ id }] });
```
Comment (lihebi) [inline ui/src/components/nodes/Scope.tsx:129]: @li-xin-yi `nodesMap.delete` and `reactFlowInstance.deleteElements` are indeed different, te latter seems to support recursive deletion of a scope.

File: ui/src/components/Canvas.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (lihebi) [reply ui/src/components/Canvas.tsx:241]: I've finished this PR. Feel free to leave comments!

[the remark this answers is not in the document]
Comment (lihebi) [reply ui/src/components/Canvas.tsx:241]: > I see, I think you can call remote deletions to DB only in `onNodesChange` here
> 
> https://github.com/codepod-io/codepod/blob/abc81df884096335175fb8cad84309f82f03d426/ui/src/lib/nodes.tsx#L79
> 
> Then all deletions observed from `nodesMap` are synced from remote peers (yes maybe from myself, but `deletePod` will check if the pod exists first).

I'm handling both zustand-store and database updates in Yjs observer (which is triggered when this user or other users make change to `nodesMap`). And I'm using `onNodesChange` only for updating nodesMap when there're reactflow operations.

**Pull request merged**